### ISSUE TITLE
ffprobe says "invalid STSD entries 0" after mp4fragment

### ISSUE THREAD

**Issue by calvincorreli** [axiomatic-systems/Bento4#245]
```
mp4fragment low.mp4 lowf.mp4

ffprobe -v error -print_format json -show_format -show_streams lowf.mp4
{
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x322df00] invalid STSD entries 0
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x322df00] error reading header
lowf.mp4: Invalid data found when processing input

}
```

I have the video (low.mp4) exhibiting the problem, but I'm not at liberty to share it publicly, so if you can reach out to me at calvin@correli.me, I'm happy to provide a URL where you can download it.

**Comment by barbibulle** [comment]
I was able to look into this.
This seems like an invalid complaint from ffprobe unfortunately.
Here's what's going on: if you use ffprobe on the original (non-fragmented) file, you'll see something like:
[...]
Unsupported codec with id 0 for input stream 2
Unsupported codec with id 0 for input stream 3
[...]
Which makes sense, since there are two 'data' tracks (ID=3 and ID=4) in the file, and those don't have a sample description (the sample description is that part of the metadata that provides information about the codec used), which isn't by itself necessarily a problem.

After you fragment the file with `mp4fragment`, the file still doesn't have a sample description fro tracks 3 and 4 (because it wasn't in the input file, `mp4fragment` can't just make one up). But this time, because it is a fragmented file, the number of samples in the 'moov' header is 0 (all the samples are in fragments, further down in the file). This triggers a different check in ffprobe, which now complains about have 0 'stsd' (sample descriptions) entries. 

So, in short, ffprobe doesn't complain about 'invalid STSD entries 0' in the non-fragmented case, because the tracks are somehow filtered out sooner for non-fragmented files. For fragmented files, ffprobe hits a different check, and complains about 'invalid STSD entries 0'.

Other players don't have a problem with the resulting file (technically, I don't see anything in the spec that specifically forbids tracks with no STSD entries), so I was able to player the fragmented file with VLC or QuickTime.
The simplest way to fix this would be to not have tracks without sample descriptions at all (maybe filter them out), so that ffprobe doesn't complain. Or don't use ffprobe, since it is checking for something that's not an error.

**Comment by barbibulle** [comment]
Actually, following up to my previous note: the reason why ffprobe complains about 'invalid STSD entries 0' for the output of the mp4fragment tool and not for the input file is that the input file, while it actually has 0 STSD entries, it has an 'stsd' box with an 'entries' field set to 1, which is incorrect, since the 'stsd' box contains 0 'SampleEntry' data structures. So the input file that ffprobe doesn't complain is actually an invalid file. When `mp4fragment` parses that input file, it does parse 'entries = 1', but since there are no entries, the array end up being empty. When the output file is written, the 'entries' field is therefore set to 0 (setting it to 1 would make the file invalid, and the Bento4 tools would never want to output an invalid file...).

**Comment by calvincorreli** [comment]
Thank you SO much for looking into this. This was everything I needed to fix this problem for me.

I'm now analyzing the input file and picking one video stream + all all audio streams, which I mix, and nothing else, and that solves it.

What do you say we close this one?

**Comment by barbibulle** [comment]
I'm glad the issue is solved. Closing this.

**Comment by eric-wieser** [comment]
This also breaks windows. I found myself in a similar situation, doing something akin to
```
mp4extract /moov/udta/ilst my.mp4 ilst.bin
mp4edit --replace /moov/udta/ilst:ilst.bin my.mp4 my-new.mp4
```

`my.mp4` works just fine, but `my-new.mp4`. Looking at a binary diff, your diagnosis above seems correct (showing `my.mp4` with the diff highlighted):

![image](https://user-images.githubusercontent.com/425260/58373608-8aa21780-7ee5-11e9-8c70-6b54b3b6d7ff.png)

What's frustrating here is that `mp4dump my.mp4` shows `stsd/entry-count` set to 0 (and gives identical output to `mp4dump my-new.mp4`).
